server/src/main/java/org/elasticsearch/health/GetHealthAction.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class GetHealthAction extends ActionType<GetHealthAction.Response> {
+
+    public static final GetHealthAction INSTANCE = new GetHealthAction();
+    // TODO: Decide on name
+    public static final String NAME = "cluster:internal/health";
+
+    private GetHealthAction() {
+        super(NAME, GetHealthAction.Response::new);
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private final ClusterName clusterName;
+        private final List<Component> components;
+
+        public Response(ClusterName clusterName, List<Component> components) {
+            this.clusterName = clusterName;
+            this.components = components;
+        }
+
+        public Response(StreamInput in) {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.startObject();
+            builder.field("status", "green");
+            builder.field("cluster_name", clusterName.value());
+            builder.array("impacts");
+            builder.startObject("components");
+            for (Component component : components) {
+                builder.field(component.getName());
+                component.toXContent(builder, params);
+            }
+            builder.endObject();
+            return builder.endObject();
+        }
+    }
+
+    public static class Request extends ActionRequest {
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    public static class TransportAction extends org.elasticsearch.action.support.TransportAction<Request, Response> {
+
+        private final ClusterService clusterService;
+
+        @Inject
+        public TransportAction(
+            final ActionFilters actionFilters,
+            final TransportService transportService,
+            final ClusterService clusterService
+        ) {
+            super(NAME, actionFilters, transportService.getTaskManager());
+            this.clusterService = clusterService;
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+            final ClusterState clusterState = clusterService.state();
+            final Controller component = new Controller(clusterService.localNode(), clusterState);
+            listener.onResponse(new Response(clusterService.getClusterName(), Arrays.asList(component, new Backups())));
+        }
+    }
+
+    public abstract static class Component implements ToXContentObject {
```

Comment:
I would like to bring this topic again:

>Looks like current both implementations are very similar. Should this be an object instead?

I believe we could have a pojo that have a final list of indicators. Logic of combining the statuses should also be the same for each component.